File: apps/web/src/entities/backlog/api/schemas.ts
import { z } from 'zod'

export const BacklogPriority = z.enum(['low', 'medium', 'high', 'critical'])
export type BacklogPriority = z.infer<typeof BacklogPriority>

export const BacklogItemCreateSchema = z.object({
  teamUuid: z.string(),
  sprintUuid: z.string(),
  title: z.string(),
  assigneeUserUuid: z.string().nullable().optional(),
  priority: BacklogPriority.optional().default('medium'),
  note: z.string().nullable().optional(),
})
export type BacklogItemCreate = z.infer<typeof BacklogItemCreateSchema>

export const BacklogItemReadSchema = z.object({
  uuid: z.string(),
  teamUuid: z.string(),
  sprintUuid: z.string(),
  title: z.string(),
  assigneeUserUuid: z.string().nullable(),
  priority: BacklogPriority,
  note: z.string().nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
})
export type BacklogItemRead = z.infer<typeof BacklogItemReadSchema>

export const BacklogItemUpdateSchema = z.object({
  title: z.string().optional(),
  assigneeUserUuid: z.string().nullable().optional(),
  priority: BacklogPriority.optional(),
  note: z.string().nullable().optional(),
  sprintUuid: z.string().optional(),
})
export type BacklogItemUpdate = z.infer<typeof BacklogItemUpdateSchema>

export const BacklogPriorityItemSchema = z.object({
  uuid: z.string(),
  priority: BacklogPriority,
})

export const BacklogPriorityUpdateSchema = z.object({
  items: z.array(BacklogPriorityItemSchema),
})
export type BacklogPriorityUpdate = z.infer<typeof BacklogPriorityUpdateSchema>

File: server/src/app/backlog/schemas.py
from datetime import datetime

from pydantic import BaseModel

from .model import BacklogPriority


class BacklogItemCreate(BaseModel):
    team_uuid: str
    sprint_uuid: str
    title: str
    assignee_user_uuid: str | None = None
    priority: BacklogPriority = BacklogPriority.MEDIUM
    note: str | None = None


class BacklogItemRead(BaseModel):
    uuid: str
    team_uuid: str
    sprint_uuid: str
    title: str
    assignee_user_uuid: str | None
    priority: BacklogPriority
    note: str | None
    created_at: datetime
    updated_at: datetime


class BacklogItemUpdate(BaseModel):
    title: str | None = None
    assignee_user_uuid: str | None = None
    priority: BacklogPriority | None = None
    note: str | None = None
    sprint_uuid: str | None = None


class BacklogPriorityItem(BaseModel):
    uuid: str
    priority: BacklogPriority


class BacklogPriorityUpdate(BaseModel):
    """Bulk reorder: list of backlog item UUIDs with their new priorities."""

    items: list[BacklogPriorityItem]
